client/src/lib.rs

```diff
@@ -0,0 +1,326 @@
+	#[fail(display = "Server returned rpc error {}", _0)]
+	JsonRpcError(Error),
+	/// Failure to parse server response.
+	#[fail(display = "Failed to parse server response")]
+	ParseError,
+	/// Request timed out.
+	#[fail(display = "Request timed out")]
+	Timeout,
+	/// The server returned a response with an unknown id.
+	#[fail(display = "Server returned a response with an unknown id")]
+	UnknownId,
+	/// Not rpc specific errors.
+	#[fail(display = "{}", _0)]
+	Other(failure::Error),
+}
+
+impl From<Error> for RpcError {
+	fn from(error: Error) -> Self {
+		RpcError::JsonRpcError(error)
+	}
+}
+
+/// The future retured by the client.
+pub struct RpcFuture {
+	recv: oneshot::Receiver<Result<Value, Error>>,
+}
+
+impl RpcFuture {
+	/// Creates a new `RpcFuture`.
+	pub fn new(recv: oneshot::Receiver<Result<Value, Error>>) -> Self {
+		RpcFuture { recv }
+	}
+}
+
+impl Future for RpcFuture {
+	type Item = Value;
+	type Error = RpcError;
+
+	fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
+		// TODO should timeout (#410)
+		match self.recv.poll() {
+			Ok(Async::Ready(Ok(value))) => Ok(Async::Ready(value)),
+			Ok(Async::Ready(Err(error))) => Err(RpcError::JsonRpcError(error)),
+			Ok(Async::NotReady) => Ok(Async::NotReady),
+			Err(error) => Err(RpcError::Other(error.into())),
+		}
+	}
+}
+
+/// A message sent to the `RpcClient`. This is public so that
+/// the derive crate can generate a client.
+pub struct RpcMessage {
+	/// The rpc method name.
+	pub method: String,
+	/// The rpc method parameters.
+	pub params: Params,
+	/// The oneshot channel to send the result of the rpc
+	/// call to.
+	pub sender: oneshot::Sender<Result<Value, Error>>,
+}
+
+/// A channel to a `RpcClient`.
+pub type RpcChannel = mpsc::Sender<RpcMessage>;
+
+/// The RpcClient handles sending and receiving asynchronous
+/// messages through an underlying transport.
+pub struct RpcClient<TSink, TStream> {
+	id: u64,
+	queue: HashMap<Id, oneshot::Sender<Result<Value, Error>>>,
+	sink: TSink,
+	stream: TStream,
+	channel: Option<mpsc::Receiver<RpcMessage>>,
+	outgoing: VecDeque<String>,
+}
+
+impl<TSink, TStream> RpcClient<TSink, TStream> {
+	/// Creates a new `RpcClient`.
+	pub fn new(sink: TSink, stream: TStream, channel: mpsc::Receiver<RpcMessage>) -> Self {
+		RpcClient {
+			id: 0,
+			queue: HashMap::new(),
+			sink,
+			stream,
+			channel: Some(channel),
+			outgoing: VecDeque::new(),
+		}
+	}
+
+	fn next_id(&mut self) -> Id {
+		let id = self.id;
+		self.id = id + 1;
+		Id::Num(id)
+	}
+}
+
+impl<TSink, TStream> Future for RpcClient<TSink, TStream>
+where
+	TSink: Sink<SinkItem = String, SinkError = RpcError>,
+	TStream: Stream<Item = String, Error = RpcError>,
+{
+	type Item = ();
+	type Error = RpcError;
+
+	fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
+		// Handle requests from the client.
+		loop {
+			if self.channel.is_none() {
+				break;
+			}
+			let msg = match self.channel.as_mut().unwrap().poll() {
```

Comment:
We prefer to use `expect("<proof>; qed")` instead of `unwrap`s. So something like `expect("channel is some; qed")`.